### PR TITLE
Counting scope vs let with async

### PR DESCRIPTION
**Pull request by jesswong**
Blurb about the differences between counting_scope and let_with_async_scope and when to use them

### CODE REVIEW

File: asyncscope.md [hunk 1]
```diff
@@ -290,6 +290,41 @@ int main() {
...
+struct work_item;
+void do_work(work_context&, work_item*);
+std::vector<work_item*> get_work_items();
+
+int main() {
+    static_thread_pool my_pool{8};
+    work_context ctx;         // create a global context for the application
+
+    std::vector<work_item*> items = get_work_items();
+    this_thread::sync_wait(ex::let_with_async_scope([&](auto scope) {
```
Comment (ispeters) [inline asyncscope.md]: @anthonywilliams is proposing that `let_with_async_scope` take a predecessor sender, like `let_value`, and his argument that it makes composition into pipelines easier makes sense to me. Perhaps this should be:
```suggestion
    this_thread::sync_wait(ex::let_with_async_scope(ex::just(get_work_items()), [&](auto scope, auto& items) {
```

File: asyncscope.md [hunk 2]
```diff
@@ -290,6 +290,41 @@ int main() {
...
+        for (auto item : items) {
+            // Spawn some work dynamically
+            ex::sender auto snd = ex::transfer_just(my_pool.get_scheduler(), item) |
+                                  ex::then([&](work_item* item) { do_work(ctx, item); }) | 
+                                  ex::let_error([](auto& e) { return just(); });
```
Comment (ispeters) [inline asyncscope.md]: We should probably call attention to this `let_error(just)` to make this pipeline compatible with spawn.
```suggestion
                                  ex::let_error([](auto& e) {
                                    // this pipeline must handle errors to be compatible with spawn(); in this
                                    // example we just swallow it, but real code could make another choice
                                    return ex::just();
                                  });
```

File: asyncscope.md [hunk 3]
```diff
@@ -345,11 +380,44 @@ int main() {
...
+            // start `snd` as before, but associate the spawned work with `scope` so that it can
+            // be awaited before destroying the resources referenced by the work (i.e. `my_pool`
+            // and `ctx`)
+            ex::spawn(std::move(snd), scope); // NEW!
+         }
+        return just();
+    }));
+
+    // `ctx` and `my_pool` are destroyed *after* they are no longer referenced
+}
```
Comment (ispeters) [inline asyncscope.md]: This looks like a duplication of the new code earlier in this diff; am I missing something?

File: asyncscope.md [hunk 4]
```diff
@@ -345,11 +380,44 @@ int main() {
...
+
+```
 :::
 
 Please see below for more examples.
 
-## `counting_scope` is step forward towards Structured Concurrency
+## `counting_scope` and `let_with_async_scope` is a step forward towards Structured Concurrency
```
Comment (ispeters) [inline asyncscope.md]:
```suggestion
## `counting_scope` and `let_with_async_scope` are a step forward towards Structured Concurrency
```

File: asyncscope.md [hunk 5]
```diff
@@ -1094,6 +1162,31 @@ requirement to complete on the receiver's scheduler restricts which receivers a
...
+Although `counting_scope` and `let_with_async_scope` have overlapping use-cases, we specifically designed the two
+facilities to address separate problems. In short, `counting_scope` is best used in an unstructured context and `let_with_async_scope`
+is best used in a structured context.
+
+We define "unstructured context" as:
+- a place where using `sync_wait` would be inappropriate, 
+- and you can't "solve by induction" (i.e you're not in an async context where you can start the sender by "awaiting" it)
+
+`counting_scope` should be used when you have a sender you want to start in an unstructured context. In this case, `spawn(sender, scope)`
+ would be the preferred way of starting asynchronous work. `scope.join()` needs to be called upon the owning object's destruction
```
Comment (ispeters) [inline asyncscope.md]:
```suggestion
 would be the preferred way of starting asynchronous work. `scope.join()` needs to be called before the owning object's destruction
```

File: asyncscope.md [hunk 6]
```diff
@@ -1094,6 +1162,31 @@ requirement to complete on the receiver's scheduler restricts which receivers a
...
+needs to be handled explicitly in the use of `counting_scope`.
+
+`let_with_async_scope` returns a sender, and therefore can only be started in one of 3 ways:
+1. `sync_wait`
+2. `spawn` on a `counting_scope`
+3. await
+
+`let_with_async_scope` will manage the scope for you which, as a result, won't give you the ability to manually manage the lifetime of
+resources the way `counting_scope` does. This behavior is a feature, since the scope being managed by `let_with_async_scope` is intended
+to live only until the sender completes. This also means that `let_with_async_scope` will be exception safe by default.
```
Comment (ispeters) [inline asyncscope.md]: > won't give you the ability to manually manage the lifetime of resources the way `counting_scope` does

I'm worried this comes off as a negative thing. Your next sentence is an explanation of why it's a feature, but I'm wondering if there's a wording that makes it sound positive from the start. I'm not sure what to suggest, but here's an attempt:
```suggestion
`let_with_async_scope` will manage the scope for you, ensuring that the managed scope is always joined before
`let_with_async_scope` completes.  The algorithm frees the user from having to manage the coupling between the
lifetimes of the managed scope and the resource(s) it protects at the cost of having to describe the nested work in
a fully structured way.
```
or
```suggestion
`let_with_async_scope` will manage the scope for you, ensuring that the managed scope is always joined before
`let_with_async_scope` completes.  The algorithm frees the user from having to manage the coupling between the
lifetimes of the managed scope and the resource(s) it protects with the limitation that the nested work must be
fully structured.
```

**Review by ispeters — COMMENTED** [review]
Looks great! Left some suggestions.

File: asyncscope.md (hunk 3 above)
Comment (jesswong) [reply asyncscope.md]: Oh it's doing the tony table comparison to follow the structure of the original version of the paper

File: asyncscope.md [hunk 7]
```diff
@@ -319,7 +356,7 @@ int main() {
 }
 ```
 
-### After
+### With counting_scope
```
Comment (ispeters) [inline asyncscope.md]: Nit
```suggestion
### With `counting_scope`
```

File: asyncscope.md [hunk 8]
```diff
@@ -348,11 +385,44 @@ int main() {
 }
 ```
 
+### With let_with_async_scope
```
Comment (ispeters) [inline asyncscope.md]: Nit
```suggestion
### With `let_with_async_scope`
```

File: asyncscope.md (hunk 3 above)
Comment (ispeters) [reply asyncscope.md]: I think in that case it needs to be shortened to match the other two examples in the table
```suggestion
```cpp
namespace ex = std::execution;

struct context;
ex::sender auto work(const context&);

int main() {
  context ctx;
  this_thread::sync_wait(ex::let_with_async_scope([&](auto scope) {
    ex::sender auto snd = work(ctx) | 
                           let_error([](auto&&) {
			    // do something to handle error
			     return just();
			  });
    // fire, but don't forget
    ex::spawn(std::move(snd), scope.get_token());
  }));

    // `ctx` is destroyed once nothing
  // references it
}
```

**Review by ispeters — COMMENTED** [review]
Looks good!

**Review by ispeters — APPROVED** [review]

**Pull request merged**